src/main/java/edu/ucla/library/iiif/auth/handlers/ItemsHandler.java

```diff
@@ -0,0 +1,116 @@
+
+import info.freelibrary.util.HTTP;
+import info.freelibrary.util.Logger;
+import info.freelibrary.util.LoggerFactory;
+
+import edu.ucla.library.iiif.auth.Error;
+import edu.ucla.library.iiif.auth.MessageCodes;
+import edu.ucla.library.iiif.auth.ResponseJsonKeys;
+import edu.ucla.library.iiif.auth.services.DatabaseService;
+import edu.ucla.library.iiif.auth.services.DatabaseServiceError;
+import edu.ucla.library.iiif.auth.services.DatabaseServiceImpl;
+import edu.ucla.library.iiif.auth.utils.MediaType;
+
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import io.vertx.core.http.HttpHeaders;
+import io.vertx.core.http.HttpServerRequest;
+import io.vertx.core.http.HttpServerResponse;
+import io.vertx.core.json.DecodeException;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.RoutingContext;
+import io.vertx.serviceproxy.ServiceException;
+
+/**
+ * Handler that handles items requests.
+ */
+public class ItemsHandler implements Handler<RoutingContext> {
+
+    /**
+     * The handler's logger.
+     */
+    private static final Logger LOGGER = LoggerFactory.getLogger(ItemsHandler.class, MessageCodes.BUNDLE);
+
+    /**
+     * The service proxy for accessing the database.
+     */
+    private final DatabaseService myDatabaseServiceProxy;
+
+    /**
+     * Creates a handler that adds items to the database.
+     *
+     * @param aVertx The Vert.x instance
+     */
+    public ItemsHandler(final Vertx aVertx) {
+        myDatabaseServiceProxy = DatabaseService.createProxy(aVertx);
+    }
+
+    @Override
+    public void handle(final RoutingContext aContext) {
+        final JsonArray requestData;
+
+        try {
+            requestData = aContext.getBodyAsJsonArray();
+        } catch (final DecodeException details) {
+            final JsonObject error = new JsonObject() //
+                    .put(ResponseJsonKeys.ERROR, Error.INVALID_JSONARRAY_ERROR) //
+                    .put(ResponseJsonKeys.MESSAGE, details.getMessage());
+
+            aContext.response().putHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON.toString())
+                    .setStatusCode(HTTP.BAD_REQUEST).end(error.encodePrettily());
+            return;
+        }
+
+        myDatabaseServiceProxy.setItems(requestData).onSuccess(result -> {
+            aContext.response().setStatusCode(HTTP.CREATED).end();
+        }).onFailure(aContext::fail);
+    }
+
+    /**
+     * Handle failure events sent by {@link #handle}.
+     *
+     * @param aContext the failure event to handle
+     */
+    public static void handleFailure(final RoutingContext aContext) {
+        final ServiceException error;
+        final HttpServerRequest request;
+        final HttpServerResponse response;
+        final String responseMessage;
+        final JsonObject data;
+        final DatabaseServiceError errorCode;
+
+        try {
+            error = (ServiceException) aContext.failure();
```

Comment:
Fwiw, I was able to send the program into an infinite loop at this point in the AbstractAccessTokenHandler (when testing with the CantaloupeDelegate). That's neither here nor there for this PR, though. I'll try to create a reproducer and file it as a bug against the project.

Is there any potential for making the errors (DatabaseServiceError/AccessCookieServiceError) extend the same base error class and consolidate this error handling code? No, is a fine answer. Just asking because I recognize the similarity of the code with that in the AbstractAccessTokenHandler.

Reply:
I noticed the infinite loop as well. The code in the catch block is incorrect, the method just keeps calling itself at L88 and never completes the response. I think it happens when the body handler fails the context before it gets to ItemsHandler, in which case the failure wouldn't be a ServiceException (and the cast throws an error).

👍🏻  on your suggestion